contracts/simple_dex/lib.rs

```diff
@@ -0,0 +1,483 @@
+        type ContractError = DexError;
+    }
+
+    impl SimpleDex {
+        #[ink(constructor)]
+        pub fn new(tokens: [AccountId; 4]) -> Self {
+            let caller = Self::env().caller();
+            let code_hash = Self::env()
+                .own_code_hash()
+                .expect("Called new on a contract with no code hash");
+            let required_role = Role::Initializer(code_hash);
+            let access_control = AccountId::from(ACCESS_CONTROL_PUBKEY);
+
+            let role_check = <Self as AccessControlled>::check_role(
+                access_control,
+                caller,
+                required_role,
+                Self::cross_contract_call_error_handler,
+                Self::access_control_error_handler,
+            );
+
+            match role_check {
+                Ok(_) => initialize_contract(|contract| Self::new_init(contract, tokens)),
+                Err(why) => panic!("Could not initialize the contract {:?}", why),
+            }
+        }
+
+        /// Swaps the a specified amount of one of the pool's PSP22 tokens to another PSP22 token
+        /// Calling account needs to give allowance to the DEX contract to spend amount_token_in of token_in on it's behalf
+        /// before executing this tx.
+        #[ink(message)]
+        pub fn swap(
+            &mut self,
+            token_in: AccountId,
+            token_out: AccountId,
+            amount_token_in: Balance,
+            min_amount_token_out: Balance,
+        ) -> Result<(), DexError> {
+            let this = self.env().account_id();
+            let caller = self.env().caller();
+
+            // check if tokens are supported by the pool
+            if !self.tokens.contains(&token_in) {
+                return Err(DexError::UnsupportedToken(token_in));
+            }
+
+            if !self.tokens.contains(&token_out) {
+                return Err(DexError::UnsupportedToken(token_out));
+            }
+
+            // check allowance
+            if self.allowance(token_in, caller, this)? < amount_token_in {
+                return Err(DexError::InsufficientAllowanceOf(token_in));
+            }
+
+            let balance_token_in = self.balance_of(token_in, this)?;
+            let balance_token_out = self.balance_of(token_out, this)?;
+
+            if balance_token_out < min_amount_token_out {
+                // throw early if we cannot support this swap anyway due to liquidity being too low
+                return Err(DexError::NotEnoughLiquidityOf(token_out));
+            }
+
+            let amount_token_out = Self::out_given_in(
+                amount_token_in,
+                balance_token_in,
+                balance_token_out,
+                self.swap_fee_percentage,
+            )?;
+
+            if balance_token_out < amount_token_out {
+                // liquidity too low
+                return Err(DexError::NotEnoughLiquidityOf(token_out));
+            }
```

Comment:
This isn't necessary - we've already checked if `balance_token_out < min_amount_token_out` and the latter has to be less than `amount_token_out`: we check that as well just below.